pageserver/src/layered_repository.rs

```diff
@@ -1300,27 +1320,53 @@ impl LayeredTimeline {
             // open layer to hold changes newer than the cutoff.) Then we write
             // the frozen layer to disk, and replace the in-memory frozen layer
             // with the new on-disk layers.
-            let FreezeLayers {
-                frozen,
-                open: maybe_new_open,
-            } = oldest_layer.freeze(last_record_lsn)?;
-
-            // replace this layer with the new layers that 'freeze' returned
-            layers.pop_oldest_open();
-            if let Some(new_open) = maybe_new_open.clone() {
-                layers.insert_open(new_open);
+            //
+            let frozen: Arc<InMemoryLayer>;
+            let maybe_new_open: Option<Arc<InMemoryLayer>>;
+
+            if !oldest_layer.is_writeable() {
+                // It is possible that the layer is already frozen (non-writeable),
+                // just skip the freeze step and write it to disk.
+                info!(
+                    "the oldest layer for {} is non-writeable. end_lsn is {}",
+                    oldest_layer.filename().display(),
+                    oldest_layer.get_end_lsn()
+                );
+                assert!(oldest_layer.is_dropped());
+                frozen = oldest_layer;
+                // There should be no newer layer that refers this non-writeable layer,
+                // because layer that is created after dropped one represents a new rel.
+                maybe_new_open = None;
+                layers.pop_oldest_open();
+            } else {
+                let FreezeLayers {
+                    frozen: frozen_result,
+                    open: open_result,
+                } = oldest_layer.freeze(last_record_lsn)?;
+
+                frozen = frozen_result;
+                maybe_new_open = open_result;
+
+                // replace this layer with the new layers that 'freeze' returned
+                layers.pop_oldest_open();
+                if let Some(new_open) = maybe_new_open.clone() {
+                    layers.insert_open(new_open);
+                }
             }
 
+            // We temporarily insert InMemory layer into historic list here.
+            // TODO: check that all possible concurrent users of 'historic' treat it right
             layers.insert_historic(frozen.clone());
 
             // Write the now-frozen layer to disk. That could take a while, so release the lock while do it
             drop(layers);
             let new_historics = frozen.write_to_disk(self)?;
-            layers = self.layers.lock().unwrap();
 
             // Finally, replace the frozen in-memory layer with the new on-disk layers
+            layers = self.layers.lock().unwrap();
             layers.remove_historic(frozen.as_ref());
 
+            //FIXME This needs a comment.
```

Comment:
@patins , could you please add a comment for this code introduced by your previous commit.
Feel free to push to this branch.